src/core/qgsfeaturefiltermodel.cpp

```diff
@@ -0,0 +1,466 @@
+    if ( mExtraIdentifierValueIndex != -1 )
+    {
+      for ( int i = 0; i < newEntriesSize; ++i )
+      {
+        if ( entries.at( i ).identifierValue == mExtraIdentifierValue )
+        {
+          currentEntryInNewList = i;
+          mEntries.replace( mExtraIdentifierValueIndex, entries.at( i ) );
+          emit dataChanged( index( mExtraIdentifierValueIndex, 0, QModelIndex() ), index( mExtraIdentifierValueIndex, 0, QModelIndex() ) );
+          setExtraValueDoesNotExist( false );
+          break;
+        }
+      }
+    }
+    else
+    {
+      Q_ASSERT_X( false, "QgsFeatureFilterModel::updateCompleter", "No extra identifier value generated. Should not get here." );
+    }
+
+    int firstRow = 0;
+
+    QgsDebugMsg( QStringLiteral( "Entries 1: %1 " ).arg( mEntries.size() ) );
+
+    // Move the extra entry to the first position
+    if ( mExtraIdentifierValueIndex != -1 )
+    {
+      if ( mExtraIdentifierValueIndex != 0 )
+      {
+        beginMoveRows( QModelIndex(), mExtraIdentifierValueIndex, mExtraIdentifierValueIndex, QModelIndex(), 0 );
+#if QT_VERSION < QT_VERSION_CHECK(5, 6, 0)
+        Entry extraEntry = mEntries.takeAt( mExtraIdentifierValueIndex );
+        mEntries.prepend( extraEntry );
+#else
+        mEntries.move( mExtraIdentifierValueIndex, 0 );
+#endif
+        endMoveRows();
+      }
+      firstRow = 1;
+    }
+
+    QgsDebugMsg( QStringLiteral( "Entries 2: %1 " ).arg( mEntries.size() ) );
+
+    // Remove all entries (except for extra entry if existent)
+    beginRemoveRows( QModelIndex(), firstRow, mEntries.size() - firstRow );
+    mEntries.remove( firstRow, mEntries.size() - firstRow );
+    endRemoveRows();
+
+    QgsDebugMsg( QStringLiteral( "Entries 3: %1 " ).arg( mEntries.size() ) );
+
+    if ( currentEntryInNewList == -1 )
+    {
+      QgsDebugMsg( QStringLiteral( "Current value is NOT in new list" ) );
+      beginInsertRows( QModelIndex(), 1, entries.size() + 1 );
+      mEntries += entries;
+      endInsertRows();
+      setExtraIdentifierValueIndex( 0 );
+    }
+    else
+    {
+      QgsDebugMsg( QStringLiteral( "Current value is in new list" ) );
+      if ( currentEntryInNewList != 0 )
+      {
+        beginInsertRows( QModelIndex(), 0, currentEntryInNewList - 1 );
+        mEntries = entries.mid( 0, currentEntryInNewList );
+        endInsertRows();
+      }
+      else
+      {
+        mEntries.replace( 0, entries.at( 0 ) );
+      }
+
+      emit dataChanged( index( currentEntryInNewList, 0, QModelIndex() ), index( currentEntryInNewList, 0, QModelIndex() ) );
+
+      beginInsertRows( QModelIndex(), currentEntryInNewList + 1, newEntriesSize - currentEntryInNewList - 1 );
+      mEntries += entries.mid( currentEntryInNewList + 1 );
+      endInsertRows();
+      setExtraIdentifierValueIndex( currentEntryInNewList );
+    }
+
+    emit filterJobCompleted();
+  }
+  emit endUpdate();
+}
+
+void QgsFeatureFilterModel::gathererThreadFinished()
+{
+  delete mGatherer;
+  mGatherer = nullptr;
```

Comment:
Do you think that `mGatherer` could be a `unique_ptr` in this case?

Reply:
I deliberately left that out. The gatherer must only be deleted after it finishes which is currently done a lot via signal/slot connections. The only time when the `unique_ptr` would actually trigger and do it's job is in the destructor. And in the destructor we pass memory management to `deleteLater`.

Reply:
OK, thank you for the explanation :)